apis/v1alpha2/config.go

```diff
@@ -0,0 +1,69 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha2
+
+import (
+	"bytes"
+	"encoding/json"
+)
+
+// Values represent parts of the config.
+type Values map[string]interface{}
+
+// DeepCopy indicate how to do a deep copy of Values type.
+func (v *Values) DeepCopy() *Values {
+	out := make(Values, len(*v))
+	for key, val := range *v {
+		switch val := val.(type) {
+		case string:
+			out[key] = val
+
+		case []string:
+			out[key] = append([]string(nil), val...)
+		default:
+			out[key] = val
+		}
+	}
+	return &out
+}
+
+// Config encapsulates collector config.
+type Config struct {
```

Comment:
The struct will have methods to simplify parsing - e.g. extract enabled component, extract pipeline name etc.

Reply:
i was wondering if we could match what the collector's config looks like right prior to the naming of various components. This would mean having explicit fields set for telemetry, pipelines, extensions, etc. ([see here](https://github.com/open-telemetry/opentelemetry-collector/blob/main/service/config.go#L15-L24)).

Something like this:
```
type Config struct {
	Receivers []receiverConfig `mapstructure:"receivers"`
	Processors []processorConfig `mapstructure:"processors"`
	Exporters []exporterConfig `mapstructure:"exporters"`
	Connectors []connectorConfig `mapstructure:"connectors"`
	Telemetry telemetryConfig `mapstructure:"telemetry"`
	Extensions extensionsConfig `mapstructure:"extensions"`
	Pipelines pipelinesConfig `mapstructure:"pipelines"`
}
```

which would let us know the type of each component when doing logic like the port parsing

Reply:
On one hand, the more structured we can make this config, the more confident we can be about rejecting invalid ones, which is a major win for users.

The potential negative is that we'll be heavily coupled to the collector at the schema level. A breaking change in the collector config schema would require us to do a corresponding breaking change in the CRD. Maybe that's ok though? Realistically, we'd need to do something like that regardless.

It would also prevent us from easily supporting alternative syntax which the collector may add without breaking changes, like https://github.com/open-telemetry/opentelemetry-collector/issues/9077.

Reply:
I like @jaronoff97's proposal. I was thinking about the same, but I wanted to do this gradually. I will play with it and add it to the current PR.

Reply:
> A breaking change in the collector config schema would require us to do a corresponding breaking change in the CRD

In the operator, we are trying to resolve breaking changes of the collector anyways. I believe that a strongly typed config will enable us to do.

Reply:
agreed with pavol. I think if / when the collector does those types of syntax changes we should have first-party support for the version that includes said changes :)